st2common/st2common/transport/connection_retry_wrapper.py

```diff
@@ -0,0 +1,164 @@
+# Licensed to the StackStorm, Inc ('StackStorm') under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import eventlet
+
+__all__ = ['ConnectionRetryWrapper', 'ClusterRetryContext']
+
+
+class ClusterRetryContext(object):
+    """
+    Stores retry context for cluster retries. It makes certain assumptions
+    on how cluster_size and retry should be determined.
+    """
+    def __init__(self, cluster_size):
+        # No of nodes in a cluster
+        self.cluster_size = cluster_size
+        # No of times to retry in a cluster
+        self.cluster_retry = 2
+        # time to wait between retry in a cluster
+        self.wait_between_cluster = 10
+
+        # No of nodes attempted. Starts at 1 since the
+        self._nodes_attempted = 1
+
+    def test_should_stop(self):
+        should_stop = True
+        if self._nodes_attempted > self.cluster_size * self.cluster_retry:
+            return should_stop, -1
+        wait = 0
+        should_stop = False
+        if self._nodes_attempted % self.cluster_size == 0:
+            wait = self.wait_between_cluster
+        self._nodes_attempted += 1
+        return should_stop, wait
+
+
+class ConnectionRetryWrapper(object):
```

Comment:
I've asked a similar thing before - can't we just use the existing failover and retry support from kombu (http://kombu.readthedocs.org/en/latest/reference/kombu.html#connection, http://kombu.readthedocs.org/en/latest/reference/kombu.html#kombu.Connection.failover_strategy, http://kombu.readthedocs.org/en/latest/reference/kombu.html#kombu.Connection.autoretry, http://kombu.readthedocs.org/en/latest/reference/kombu.html#kombu.Connection.ensure_connection)?

It looks like it already implements all the things we need (including retrying of a particular function), or am I missing something? :)

And the whole failover and retrying happening inside kombu seems better anyway since there it's closer to the place where the actual connection is established, etc. (not to mention the fact their retry code is more battle tested than ours).